src/Microsoft.TestPlatform.ObjectModel/DirectoryBasedDiscovererAttribute.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+/// <summary>
+/// This attribute is applied to <see cref="ITestDiscoverer"/>s. It indicates the test discoverer discovers tests
+/// present inside a directory (as opposed to the <see cref="FileExtensionAttribute"/> which indicates that the
+/// discoverer discovers tests present in files with a specified extension).
+/// </summary>
+/// <remarks>
+/// If neither this attribute nor the <see cref="FileExtensionAttribute"/> is provided on the test discoverer,
+/// it will be called for all relevant test files and directories.
+/// </remarks>
+[AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
+public sealed class DirectoryBasedTestDiscovererAttribute : Attribute
```

Comment:
We probably already thought of this but what is our recommendation for adapter authors that use this and are worried about compat with older VS/vstest versions that might not have this attribute defined in the object model?

Reply:
That's a good point @AbhitejJohn. This would not an issue for the python adapter specifically since it ships as part of VS (VSIX-based).

Otherwise, my assumption was that the attribute will be ignored by older versions of TP (which don't know about its existence and therefore won't look for it / try to reflection load it). I will try and test this out.

Reply:
Adapter authors also have the choice to leave out this attribute entirely. In that case, the adapter would be invoked for all sources (not just directories), and it would be up to the adapter to filter out directories in the supplied list...

Reply:
@nohwnd does it work with all the "above" part related to inferring the host?

Reply:
The compatibility is probably the same as for any other adapter, once you use this latest version of ObjectModel and use any new feature from it, you won't be able to load into older version of VS for .NET Framework tests. (I am writing a proposal to mitigate this.).

Reply:
Luckily the python adapter ships with VS (as shyam said above, so the version is fixed for .NET Framework, and can be updated for .NET Core by updating the version of Test.SDK.

Reply:
Sounds good, the only other thing I can think of is graceful failure for adapters that do use new functionality but on older versions that do not support it. At the moment, I think they wouldn't even load failing to discover or execute any tests but perhaps @nohwnd's proposal would discuss what our recommendation there would be.